### PR TITLE
Promote Style to public API

### PR DESCRIPTION
**Pull request by jfirebaugh**
Make `style.hpp` a public header, provide `Style& Map::getStyle()`, and move runtime-styling APIs that previously lived on `Map` to `Style`. This is a refactor made relatively easy by the async rendering changes. It reduces the responsibilities of the `Map` object itself and is a prerequisite for #6386. @1ec5, can you review what else is needed from core to support #6386? In particular, do we need `Map::setStyle(std::unique_ptr<Style>)`? Does `Style` need to be constructible without the `Scheduler&, FileSource&, float pixelRatio` arguments? I think we can hide those if need be.

### CODE REVIEW

File: include/mbgl/map/map.hpp [hunk 1]
```diff
@@ -58,15 +55,15 @@ class Map : private util::noncopyable {
     // Main render function.
     void render(View&);
 
-    // Styling
-    style::TransitionOptions getTransitionOptions() const;
-    void setTransitionOptions(const style::TransitionOptions&);
-
+    // Style
     void setStyleURL(const std::string&);
```
Comment (1ec5) [inline include/mbgl/map/map.hpp]: With this method, the workflow is to create the map, then set its style to a URL. #6386 also proposes exposing an  `-[MGLStyle initWithURL:]`. I think this would require a way to initialize an `mbgl::style::Style` with nothing but a URL, so that later on `mbgl::Map::setStyle()` could be called, passing in that style. Do you see any complications with that approach? `-[MGLStyle initWithJSONData:]` (analogous to `setStyleJSON()`) is also a highly requested feature.

File: platform/ios/src/MGLMapView.mm [hunk 2]
```diff
@@ -2152,10 +2153,10 @@ - (void)resetNorthAnimated:(BOOL)animated
...
 - (void)resetPosition
 {
-    CGFloat pitch = _mbglMap->getDefaultPitch();
-    CLLocationDirection heading = mbgl::util::wrap(_mbglMap->getDefaultBearing(), 0., 360.);
-    CLLocationDistance distance = MGLAltitudeForZoomLevel(_mbglMap->getDefaultZoom(), pitch, 0, self.frame.size);
-    self.camera = [MGLMapCamera cameraLookingAtCenterCoordinate:MGLLocationCoordinate2DFromLatLng(_mbglMap->getDefaultLatLng())
+    CGFloat pitch = _mbglMap->getStyle().getDefaultPitch();
+    CLLocationDirection heading = mbgl::util::wrap(_mbglMap->getStyle().getDefaultBearing(), 0., 360.);
+    CLLocationDistance distance = MGLAltitudeForZoomLevel(_mbglMap->getStyle().getDefaultZoom(), pitch, 0, self.frame.size);
+    self.camera = [MGLMapCamera cameraLookingAtCenterCoordinate:MGLLocationCoordinate2DFromLatLng(_mbglMap->getStyle().getDefaultLatLng())
```
Comment (1ec5) [inline platform/ios/src/MGLMapView.mm]: Ultimately, we’d add more internal properties to MGLStyle so that calls to `_mbglMap->getStyle()` become largely unnecessary in MGLMapView’s implementation.

**Comment by jfirebaugh** [comment]
Cool -- I moved the JSON and URL loading methods from `Map` to `Style` as well, and added `Map::setStyle`.

> I’m less sure about `Scheduler` – it would be great if we could hide that parameter.

Now that the loading methods are on `Style`, these parameters are all necessary. SDKs can pass the same `Scheduler&` as they pass to the `Map` constructor.

File: src/mbgl/map/map.cpp [hunk 3]
```diff
@@ -166,14 +153,12 @@ Map::Impl::Impl(Map& map_,
           }
       }) {
     style = std::make_unique<Style>(scheduler, fileSource, pixelRatio);
```
Comment (ivovandongen) [inline src/mbgl/map/map.cpp]: Not part of this pr. But why aren't we initialising this in the member initialisation list?

File: include/mbgl/style/style.hpp [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+    double getDefaultZoom() const;
+    double getDefaultBearing() const;
+    double getDefaultPitch() const;
+
+    // TransitionOptions
+    TransitionOptions getTransitionOptions() const;
+    void setTransitionOptions(const TransitionOptions&);
+
+    // Light
+          Light* getLight();
```
Comment (ivovandongen) [inline include/mbgl/style/style.hpp]: Should we use indentation here that is more common? Not sure if this helps with readability, but it prevents using standard formatting tools.

File: src/mbgl/style/style_impl.cpp [hunk 5]
```diff
@@ -0,0 +1,372 @@
...
+    parse(json_);
+}
+
+void Style::Impl::loadURL(const std::string& url_) {
+    observer->onStyleLoading();
+
+    loaded = false;
+    url = url_;
+
+    styleRequest = nullptr;
```
Comment (ivovandongen) [inline src/mbgl/style/style_impl.cpp]: Is this necessary? The next line would already destruct any previous style before assignment right?

File: src/mbgl/style/style_impl.cpp [hunk 6]
```diff
@@ -0,0 +1,372 @@
...
+    auto wrappers = layers.getWrappers();
+    return std::vector<const Layer*>(wrappers.begin(), wrappers.end());
+}
+
+Layer* Style::Impl::getLayer(const std::string& id) const {
+    return layers.get(id);
+}
+
+Layer* Style::Impl::addLayer(std::unique_ptr<Layer> layer, optional<std::string> before) {
+    // TODO: verify source
```
Comment (ivovandongen) [inline src/mbgl/style/style_impl.cpp]: Great you got rid of BackendScope here!

File: src/mbgl/map/map.cpp (hunk 3 above)
Comment (jfirebaugh) [inline src/mbgl/map/map.cpp]: It probably should be.

File: include/mbgl/style/style.hpp (hunk 4 above)
Comment (jfirebaugh) [inline include/mbgl/style/style.hpp]: It make clear the parallelism with the `const` version of the accessor -- you can immediately see where the `const`s are added, that they are in the right place, and that there aren't any other differences in the signatures. I'm opposed to using formatting tools that can't handle nuances like this.

File: src/mbgl/style/style_impl.cpp (hunk 5 above)
Comment (jfirebaugh) [inline src/mbgl/style/style_impl.cpp]: It's not necessary, I'll remove it.
